src/js/UnifileService.js

```diff
@@ -185,37 +186,57 @@ export default class UnifileService {
-      // Here we may not have the service info yet, e.g. if we did not ls '/'
-      if (service && service.isLoggedIn) {
-        this.authEnded(serviceName, resolve, reject);
-      } else {
-        // Open a blank window right away, before we know the URL, otherwise the browser blocks it
-        const win = window.open();
-        const req = new XMLHttpRequest();
-        req.open('POST', `${UnifileService.ROOT_URL}${serviceName}/authorize`);
-        req.onload = () => {
-          if (req.responseText) {
-            win.location = req.responseText;
-            win.addEventListener('unload', () => {
-              win.onunload = null;
-              this.startPollingAuthWin({
+      //To avoid repeat code, this function prepare the login/auth process
+  function prepLogin(win, req, resolve, reject, UnifileService, serviceName, obj){
+ 			req.open('POST', `${UnifileService.ROOT_URL}${serviceName}/authorize`);
+			req.onload = () => {
+			  if (req.responseText) {
+          win.location = req.responseText;
+          win.addEventListener('unload', () => {
+                win.onunload = null;
+                obj.startPollingAuthWin({
                 reject,
                 resolve,
                 serviceName,
                 win
-              });
-            });
-          } else {
-            this.authEnded(serviceName, resolve, reject);
+				      });
+				    });
+ 			  } else {
+            obj.authEnded(serviceName, resolve, reject);
             win.close();
-          }
-        };
-        req.onerror = reject;
-        req.ontimeout = () => reject(new Error('Auth request timed out'));
-        req.send();
-      }
+				  }			
+				}
+			  req.onerror = reject;
+			  req.ontimeout = () => reject(new Error('Auth request timed out'));
+			  req.send();
+			  return;
+     	}    
+return new Promise((resolve, reject) => {
+      const service = this.constructor.getServiceByName(serviceName);
+      // Here we may not have the service info yet, e.g. if we did not ls '/'
+      if (service && service.isLoggedIn) {
+            this.authEnded(serviceName, resolve, reject);
+      } else {
+            //For OAuth (github, droopbox...) we use a blank window, as the user is exiting from the site, so it's more evident
+            //Non Oauth, like ftp, sftp uses the modal window. I tyr to get the  minimal modification in the process
+           if (service.isOAuth){
+                const win = window.open();
+                const req = new XMLHttpRequest();
+                prepLogin(win,req,resolve, reject, UnifileService, serviceName, this);
+                }
+            else{
+                var cer = ce.cloudExplorer;				
+                var m = ModalDialog.getInstance();
+                //Change state and callback function for login in win popup
+                //Cannot have an synchronous render so need to use callback.
+                m.setState({mode:'login'},
```

Comment:
I'm not sure I fully understand, but I think you need to split this in two here :
* just setState({}) here
* move the rest of the logic to the `render()` function (`if(this.state.mode === 'login')`)

Reply:
Thanks for your review.
My idea was that setState({mode:'login') throws a render of modaldialog, that check the mode is login and shows the window popup. Indeed in modaldialog the login of check state is working now.
I do not know how to use just setState({}), how tell modaldialgo the state is login?
So I cannot modify this code.

Reply:
Ok I guess it's too much of a rework.
I really dislike that you need to `ModalDialog.getInstance()` here (I think `getInstance()` is really an anti-pattern here...), but I have no clue as to how to avoid this :S